### ISSUE TITLE
Parse fuzzy scrutiny dates

### ISSUE THREAD

**Issue by mdoering** [Sp2000/colplus-backend#37]
The latest taxonomic scrutiny date data found in the CoL is treated there as text and has various forms of partial dates.

We would like to parse all these into a structured form, either a LocalDate as we have now in the data model Taxon.accordingToDate or a new FuzzyDate class that:

- has 3 year, month, day int fields
- requires a year, but allows fuzzy dates by having a null day or even a unknown month (in that case also unknown day)
- offers a [getTemporalAccessor()](https://docs.oracle.com/javase/8/docs/api/java/time/temporal/TemporalAccessor.html) method that returns either a Year, YearMonth or LocalDate instance to interact with the java API.

If the LocalDate class is used we need to point a year to the 1st January of that year and likewise for months only. This is not ideal but an ok start.

**Comment by mdoering** [comment]
see https://github.com/Sp2000/colplus-backend/blob/master/colplus-parser/src/test/resources/dates.txt with 100 example values

**Comment by mdoering** [comment]
@ayco-at-naturalis any preference yet?

**Comment by ayco-at-naturalis** [comment]
Oops only now see your question ... again I have opted for the "simple" solution b/c it hooks most easily into existing parsing code ... FuzzyDate or something is something much grander that would necessitate a complete rethink of date parsing (also for DwCA)

**Comment by mdoering** [comment]
So we would have multiple date parsers for a LocalDate and apply them by context?
I am not sure if that is feasable, ultimately we will move some resources away from ACEF into DwCA. We are also likely facing similar dates with dwc terms. I think we better have just a single date parser.

**Comment by mdoering** [comment]
if we can get this new parser to handle all dates we encounter in other dwc or acef date fields then lets use that instead of the GBIF wrapped one.

**Comment by mdoering** [comment]
btw, here are gbif date parser and tests:
https://github.com/gbif/parsers/tree/master/src/main/java/org/gbif/common/parsers/date
https://github.com/gbif/parsers/tree/master/src/test/java/org/gbif/common/parsers/date

**Comment by mdoering** [comment]
maybe we should inventory all distinct date types we see in our data

**Comment by ayco-at-naturalis** [comment]
Yes, I think we should make an inventory of all formats and then sequentially apply those (long to short, or in some configurable sequence to avoid wasting time on improbably formats given some data source) to any date string to be parsed

**Comment by mdoering** [comment]
Looking at the data that doesnt seem too difficult, we only have very few terms which are dates.
I am dumping all distinct values for these:

- ACEF:LTSDate
- dwc:namePublishedInYear
- dc:modified

I cannot get this one from the db, please check the archives or assembly db for data:
- ACEF SourceDatabase.txt ReleaseDate

Thats all there is apart from References.
But thats a special beast as its embedded in the CSL and we can deal with that later I reckon.

So maybe your AcefDateParser can be extended and become THE date parser, ditching the GBIF beast

**Comment by ayco-at-naturalis** [comment]
Just writing ... yes ... I think the current gbif date parsing library is difficult to hook into if another date format pops up in the data, and probably more complex than needed ... we just need a list of "allowed date formates" and sequence over it.

**Comment by mdoering** [comment]
[dates.tar.gz](https://github.com/Sp2000/colplus-backend/files/1811207/date.tar.gz)

**Comment by mdoering** [comment]
ideally all dates are ISO dates, but we need to adapt to whatever we see

**Comment by mdoering** [comment]
excellent!
Before closing this we need to be sure we really want "2016-03-01" in our API instead of "March 2016". The artificially created precision is misleading. For now its sufficient, but we should keep an issue open (maybe new one that links to this) to improve the fuzzy date storage

**Comment by mdoering** [comment]
Some imports now fail with IllegalArgumentException
```Caused by: java.lang.IllegalArgumentException: Unsupported date: "Dec-2016"
	at org.col.admin.task.importer.acef.AcefDateParser.parse(AcefDateParser.java:48)
```

The parsers should never throw any other runtime exception but an UnparsableException

**Comment by mdoering** [comment]
btw, the new DateParser should live in the parser module

**Comment by ayco-at-naturalis** [comment]
I am implementing the "fuzzy date" logic on a different branch.  What's left in the master branch should probably restored to the original code until I'm ready to merge the branches (although that will cause import problems, because that's why we have this issue)

**Comment by ayco-at-naturalis** [comment]
I wonder whether a one-size-fits-all parsing (as currently happens) is a good thing. Data quality would probably improve if we would/could make the date parser data source-aware, maybe even field-aware. For example so that we can successfully (or more easily) parse dwc published-in dates like "2007a". Also we **know** that some data sources use american dates (mdy) while other use european dates (dmy), so it doesn't make sense to check for ambiguity (and it may actually be bad). 

In other words, it's probably a good thing if "inserters" and "interpreters" can configure their own parser, per data source type (ACEF, DWCA) and possibly even per field.

At least it probably pays the make date parsing more configurable.

**Comment by mdoering** [comment]
thats true, I just find it hard to know how to configure it. Maybe passing at least a preferred & expected date format (through static builder pattern?) that is tried before any of the others would be good

**Comment by ayco-at-naturalis** [comment]
Finished but not merged back to master
